plugins/quantum_k_nearest_neighbours/backend/qknns/simpleQkNN.py

```diff
@@ -0,0 +1,551 @@
+            self.point_num_to_idx[bitlist_to_int(vec)] = i
+
+        self.unclean_wires = [] if unclean_wires is None else unclean_wires
+        self.train_wires = train_wires
+        self.qam_ancilla_wires = qam_ancilla_wires
+        wire_types = ["train", "qam_ancilla", "unclean"]
+        num_wires = [self.train_data.shape[1], max(self.train_data.shape[1], 2)]
+        error_msgs = [
+            "the points' dimensionality.",
+            "the points' dimensionality and greater or equal to 2.",
+        ]
+        check_wires_uniqueness(self, wire_types)
+        check_num_wires(self, wire_types[:-1], num_wires, error_msgs)
+        self.qam = QAM(
+            self.train_data,
+            self.train_wires,
+            self.qam_ancilla_wires,
+            unclean_wires=self.unclean_wires,
+        )
+
+    def get_quantum_circuit(self, x: np.ndarray):
+        """
+        Given a binary test vector of size m. This function returns a quantum circuit that does the following:
+        1. Load data into the trainings register with a quantum associative memory
+        2. Invert the i'th qubit in the trainings register, if the i'th entry of the test vector is 0
+        => The sum of the trainings register is equal to m minus the hamming distance to the test vector.
+        3. For each trainings qubit equal to |1>, an oracle qubit is rotated by pi/m further towards |1>
+        4. Uncompute step 2
+        5. Return measurements of the trainings register and the oracle qubit
+        => The probability of the oracle qubit being |0> is proportional to the hamming distance
+        """
+        check_binary(
+            x, "All the data needs to be binary, when dealing with the hamming distance"
+        )
+        rot_angle = np.pi / len(x)
+
+        def quantum_circuit():
+            self.qam.circuit()
+            for x_, train_wire in zip(x, self.train_wires):
+                if x_ == 0:
+                    qml.PauliX((train_wire,))
+            for train_wire in self.train_wires:
+                # QAM ancilla wires are 0 after QAM -> use one of those wires
+                qml.CRX(rot_angle, wires=(train_wire, self.qam_ancilla_wires[0]))
+            for x_, train_wire in zip(x, self.train_wires):
+                if x_ == 0:
+                    qml.PauliX((train_wire,))
+            return qml.sample(wires=self.train_wires + [self.qam_ancilla_wires[0]])
+
+        return quantum_circuit
+
+    def calculate_distances(self, x: np.ndarray) -> List[float]:
+        samples = qml.QNode(self.get_quantum_circuit(x), self.backend)().tolist()
+        num_zero_ancilla = [0] * len(self.train_data)
+        total_ancilla = [0] * len(self.train_data)
+        # Count how often a certain point was measured (total_ancilla)
+        # and how often the ancilla qubit was zero (num_zero_ancilla)
+        for sample in samples:
+            idx = self.point_num_to_idx[bitlist_to_int(sample[:-1])]
```

Comment:
I know this was long ago, but do you remember using a classical simulator or a real quantum computer?

On a second note, do you remember what your dataset looks like? It seems like the same point is in the dataset more than once. This shouldn't hinder the execution, but raises some concerns about how to interpret the solution. Is it realistic to have the same point appear more than once in a dataset, potentially with different labels?

Reply:
I only used simulators.

The data looks like this:
`[{"ID": 0, "href": "", "x": 1, "y": 0}, {"ID": 1, "href": "", "x": 1, "y": 0}, {"ID": 2, "href": "", "x": 0, "y": 0}, ...`
and the labels look like this:
`[{"ID": 0, "href": "", "label": 1}, {"ID": 1, "href": "", "label": 1}, {"ID": 2, "href": "", "label": 0}, ...`

The data is randomly generated. Because it's binary and low dimensional, the chance is high that the same point appears multiple times. With real data the same thing could happen if the data is low dimensional as well. Because real data is noisy and erroneous, the same point could even have different labels.

Reply:
For now I simply added to the description  (of the affected QkNNs) that they can only process datasets that may only contain a point at most once.
The algorithms can be adapted to even work with datasets in which the points get repeated. This may be done in the future.

Reply:
This is a serious limitation that is easily overlooked. It would be better to check this property of the input values in code and error if duplicate points were found (or remove duplicate points and only mention them in the logs). If you do not update the algorithms to work with duplicate points, then add an issue for this (and add any pointers as to how the algorithms need to be updated in the issue if you already have them).